隐私设置（v1）& 关注 & 动态 & 动态分组 & 黑名单

cfdxkk opened the pull request:
开发并交付关注、动态与动态分组功能。

```
用户可以关注或取关其他用户
用户可以查看关注的用户的动态
用户可以创建分组
用户可以将用户加入或移出分组
用户可以只查看某个分组内用户的动态，而不是查看全部用户的动态（可以参考推特的“列表”功能）
```

开发并交付黑名单相关内容

```
用户可以屏蔽/隐藏（取消屏蔽/显示）其他用户
用户可以屏蔽（取消屏蔽）标签
用户可以屏蔽（取消屏蔽）关键词
用户可以添加（删除）正则表达式
用户可以获取黑名单列表
```

cfdxkk committed:
✨ 用户关注另一用户

cfdxkk committed:
✨ 1. 用户的信息可见性设置和关联平台可见性设置。2. 关注和取消关注的接口。

1. 用户的信息可见性设置和关联平台可见性设置。
2. 关注和取消关注的接口。（该接口未经任何测试）

cfdxkk committed:
✨ 获取用户信息时增加用户是否已经被关注的属性

1. 获取用户信息时可以显示用户是否已经被关注
2. 其他 QoL 改进

cfdxkk committed:
🐛 修复一个无法启动开发环境的问题。

enum 数据不应该放到 `.d.ts` 文件里，因为 tsc 构建后根本没有保留 `.d.ts` 文件
所以所有 `*Dto.d.ts` 全改为 `*Dto.ts` 了

cfdxkk committed:
🐛 改用 Pipeline，修复了首页视频偶尔不显示上传者昵称的问题

cfdxkk committed:
✨ 视频页可以获取是否已关注上传者，以及上传者是否是自己

1. 视频页可以获取是否已关注上传者，以及上传者是否是自己
2. 重写更新或创建浏览历史记录的接口，使其适配 uuid

cfdxkk committed:
🐛 修复可以多次关注一个用户的 bug，并将关注接口的请求参数从 uuid 改为 uid

cfdxkk committed:
✨ 对于角色校验失败的用户，在控制台输出警告而不是错误。

cfdxkk committed:
:art: QoL Update

cfdxkk committed:
✨ 取关接口

cfdxkk committed:
Merge remote-tracking branch 'origin/develop' into feature-20250105-Feed

cfdxkk committed:
✨ 动态分组相关 API (In progress)

cfdxkk committed:
🐛 修复一个 Merge 问题

cfdxkk committed:
✨ 删除动态分组的接口

cfdxkk committed:
✨ 动态分组相关 API（更新信息、删除、管理员审核和管理员删除）

cfdxkk committed:
✨ 获取动态分组

cfdxkk committed:
✨ 🍻 获取动态内容的接口 | 🎵 红色高跟鞋 cover 迦乐

cfdxkk committed:
🎨 Fix Typo

cfdxkk committed:
🎨 QoL Update.

cfdxkk committed:
✨ QoL Update - 更新数据时会同步更新 “最后编辑时间”

cfdxkk committed:
🐛 Fix typo

cfdxkk committed:
🐛 修复一个搜索不到动态数据时 totalCount 可能会报错的问题。

YasaitoriOvO commented:
:art: 重命名封禁相关的函数
```diff
diff --git a/src/controller/UserController.ts b/src/controller/UserController.ts
index 84e7de1..81370ee 100644
--- a/src/controller/UserController.ts
+++ b/src/controller/UserController.ts
@@ -3,13 +3,13 @@ import {
 	adminClearUserInfoService,
 	adminGetUserInfoService,
 	approveUserInfoService,
-	blockUserByUIDService,
+	banUserByUIDService,
 	changePasswordService,
 	checkInvitationCodeService,
 	checkUsernameService,
 	checkUserTokenService,
 	createInvitationCodeService,
-	getBlockedUserService,
+	getBannedUserService,
 	getMyInvitationCodeService,
 	getSelfUserInfoService,
 	getUserAvatarUploadSignedUrlService,
@@ -43,7 +43,7 @@ import {
 	AdminClearUserInfoRequestDto,
 	AdminGetUserInfoRequestDto,
 	ApproveUserInfoRequestDto,
-	BlockUserByUIDRequestDto,
+	BanUserByUIDRequestDto,
 	CheckInvitationCodeRequestDto,
 	CheckUserHave2FARequestDto,
 	CheckUsernameRequestDto,
@@ -688,16 +688,16 @@ export const checkUsernameController = async (ctx: koaCtx, next: koaNext) => {
  * @param next context
  * @return 封禁用户的请求响应
  */
-export const blockUserByUIDController = async (ctx: koaCtx, next: koaNext) => {
-	const data = ctx.request.body as Partial<BlockUserByUIDRequestDto>
-	const blockUserByUIDRequest: BlockUserByUIDRequestDto = {
+export const banUserByUIDController = async (ctx: koaCtx, next: koaNext) => {
+	const data = ctx.request.body as Partial<BanUserByUIDRequestDto>
+	const banUserByUIDRequest: BanUserByUIDRequestDto = {
 		criminalUid: data.criminalUid ?? -1,
 	}
 	const uid = parseInt(ctx.cookies.get('uid'), 10)
 	const token = ctx.cookies.get('token')
 
-	const blockUserByUIDResponse = await blockUserByUIDService(blockUserByUIDRequest, uid, token)
-	ctx.body = blockUserByUIDResponse
+	const banUserByUIDResponse = await banUserByUIDService(banUserByUIDRequest, uid, token)
+	ctx.body = banUserByUIDResponse
 	await next()
 }
 
@@ -726,11 +726,11 @@ export const reactivateUserByUIDController = async (ctx: koaCtx, next: koaNext)
  * @param next context
  * @return 获取所有被封禁用户的信息的请求响应
  */
-export const getBlockedUserController = async (ctx: koaCtx, next: koaNext) => {
+export const getBannedUserController = async (ctx: koaCtx, next: koaNext) => {
 	const uid = parseInt(ctx.cookies.get('uid'), 10)
 	const token = ctx.cookies.get('token')
 
-	const reactivateUserByUIDResponse = await getBlockedUserService(uid, token)
+	const reactivateUserByUIDResponse = await getBannedUserService(uid, token)
 	ctx.body = reactivateUserByUIDResponse
 	await next()
 }
diff --git a/src/controller/UserControllerDto.ts b/src/controller/UserControllerDto.ts
index 8f9292e..f3bffb2 100644
--- a/src/controller/UserControllerDto.ts
+++ b/src/controller/UserControllerDto.ts
@@ -309,7 +309,7 @@ export type GetUserInfoByUidResponseDto = {
 		role?: string;
 		/** 是否正在关注该用户 */
 		isFollowing: boolean;
-		/** 
+		/**
 		 * 查询的用户是否是自己。
 		 * 如果该字段的值为 true，则通常意味着发生了错误的请求，因为有专用的接口用于查询用户自己的信息。
 		 */
@@ -666,7 +666,7 @@ export type CheckUsernameResponseDto = {
 /**
  * 封锁一个用户的请求载荷
  */
-export type BlockUserByUIDRequestDto = {
+export type BanUserByUIDRequestDto = {
 	/** 将被封锁的用户的 UID */
 	criminalUid: number;
 }
@@ -674,7 +674,7 @@ export type BlockUserByUIDRequestDto = {
 /**
  * 封锁一个用户的请求响应
  */
-export type BlockUserByUIDResponseDto = {
+export type BanUserByUIDResponseDto = {
 	/** 执行结果，是否封锁成功 */
 	success: boolean;
 	/** 附加的文本消息 */
@@ -702,7 +702,7 @@ export type ReactivateUserByUIDResponseDto = {
 /**
  * 获取所有被封禁用户的信息的请求响应
  */
-export type GetBlockedUserResponseDto = {
+export type GetBannedUserResponseDto = {
 	/** 执行结果 */
 	success: boolean;
 	/** 附加的文本消息 */
diff --git a/src/route/router.ts b/src/route/router.ts
index 578bc1d..985c239 100644
--- a/src/route/router.ts
+++ b/src/route/router.ts
@@ -7,12 +7,12 @@ import {
 	adminClearUserInfoController,
 	adminGetUserInfoController,
 	approveUserInfoController,
-	blockUserByUIDController,
+	banUserByUIDController,
 	checkInvitationCodeController,
 	checkUsernameController,
 	checkUserTokenController,
 	createInvitationCodeController,
-	getBlockedUserController,
+	getBannedUserController,
 	getMyInvitationCodeController,
 	getSelfUserInfoController,
 	getUserAvatarUploadSignedUrlController,
@@ -271,8 +271,8 @@ router.post('/user/update/password', updateUserPasswordController) // 更新用
 router.get('/user/checkUsername', checkUsernameController) // 检查用户名是否可用
 // https://localhost:9999/user/checkUsername?username=xxxxxxxx
 
-router.post('/user/blockUser', blockUserByUIDController) // 根据 UID 封禁一个用户 // WARN: 仅限管理员
-// https://localhost:9999/user/blockUser
+router.post('/user/banUser', banUserByUIDController) // 根据 UID 封禁一个用户 // WARN: 仅限管理员
+// https://localhost:9999/user/banUser
 // cookie: uid, token
 // {
 // 	"criminalUid": XXXX
@@ -285,8 +285,8 @@ router.post('/user/reactivateUser', reactivateUserByUIDController) // 根据 UID
 // 	"uid": XXXX
 // }
 
-router.get('/user/blocked/info', getBlockedUserController) // 获取所有被封禁用户的信息 // WARN: 仅限管理员
-// https://localhost:9999/user/blocked/info
+router.get('/user/baned/info', getBannedUserController) // 获取所有被封禁用户的信息 // WARN: 仅限管理员
+// https://localhost:9999/user/baned/info
 // cookie: uid, token
 
 router.get('/user/adminGetUserInfo', adminGetUserInfoController) // 管理员获取用户信息 // WARN: 仅限管理员
diff --git a/src/service/UserService.ts b/src/service/UserService.ts
index 3474360..3e9b5d7 100644
--- a/src/service/UserService.ts
+++ b/src/service/UserService.ts
@@ -11,8 +11,8 @@ import {
 	AdminGetUserInfoResponseDto,
 	ApproveUserInfoRequestDto,
 	ApproveUserInfoResponseDto,
-	BlockUserByUIDRequestDto,
-	BlockUserByUIDResponseDto,
+	BanUserByUIDRequestDto,
+	BanUserByUIDResponseDto,
 	CheckInvitationCodeRequestDto,
 	CheckInvitationCodeResponseDto,
 	CheckUsernameRequestDto,
@@ -20,7 +20,7 @@ import {
 	CheckUserTokenResponseDto,
 	CreateInvitationCodeResponseDto,
 	DeleteTotpAuthenticatorByTotpVerificationCodeResponseDto,
-	GetBlockedUserResponseDto,
+	GetBannedUserResponseDto,
 	GetMyInvitationCodeResponseDto,
 	GetSelfUserInfoRequestDto,
 	GetSelfUserInfoResponseDto,
@@ -702,7 +702,7 @@ export const updateUserEmailService = async (updateUserEmailRequest: UpdateUserE
  */
 export const updateOrCreateUserInfoService = async (updateOrCreateUserInfoRequest: UpdateOrCreateUserInfoRequestDto, uid: number, token: string): Promise<UpdateOrCreateUserInfoResponseDto> => {
 	try {
-		if (await checkUserRoleService(uid, 'blocked')) {
+		if (await checkUserRoleService(uid, 'baned')) {
 			console.error('ERROR', '更新或创建用户信息失败，用户已封禁')
 			return { success: false, message: '更新或创建用户信息失败，用户已封禁' }
 		}
@@ -2393,31 +2393,31 @@ export const checkUserExistsByUuidService = async (checkUserExistsByUuidRequest:
 
 /**
  * 根据 UID 封禁一个用户
- * @param blockUserByUIDRequest 封禁用户的请求载荷
+ * @param banUserByUIDRequest 封禁用户的请求载荷
  * @param adminUid 管理员的 UID
  * @param adminToken 管理员的 Token
  * @returns 封禁用户的请求响应
  */
-export const blockUserByUIDService = async (blockUserByUIDRequest: BlockUserByUIDRequestDto, adminUid: number, adminToken: string): Promise<BlockUserByUIDResponseDto> => {
+export const banUserByUIDService = async (banUserByUIDRequest: BanUserByUIDRequestDto, adminUid: number, adminToken: string): Promise<BanUserByUIDResponseDto> => {
 	try {
-		if (checkBlockUserByUIDRequest(blockUserByUIDRequest)) {
+		if (checkBanUserByUIDRequest(banUserByUIDRequest)) {
 			if (await checkUserToken(adminUid, adminToken)) {
 				const isAdmin = await checkUserRoleService(adminUid, 'admin')
 				if (isAdmin) {
-					const { criminalUid } = blockUserByUIDRequest
+					const { criminalUid } = banUserByUIDRequest
 					const { collectionName, schemaInstance } = UserAuthSchema
 					type UserAuth = InferSchemaType<typeof schemaInstance>
 
-					const blockUserByUIDWhere: QueryType<UserAuth> = {
+					const banUserByUIDWhere: QueryType<UserAuth> = {
 						uid: criminalUid,
 						role: 'user',
 					}
 
-					const blockUserByUIDUpdate: UpdateType<UserAuth> = {
-						role: 'blocked',
+					const banUserByUIDUpdate: UpdateType<UserAuth> = {
+						role: 'baned',
 					}
 					try {
-						const updateResult = await findOneAndUpdateData4MongoDB<UserAuth>(blockUserByUIDWhere, blockUserByUIDUpdate, schemaInstance, collectionName, undefined, false)
+						const updateResult = await findOneAndUpdateData4MongoDB<UserAuth>(banUserByUIDWhere, banUserByUIDUpdate, schemaInstance, collectionName, undefined, false)
 						if (updateResult.success && updateResult.result) {
 							return { success: true, message: '封禁用户成功' }
 						} else {
@@ -2465,7 +2465,7 @@ export const reactivateUserByUIDService = async (reactivateUserByUIDRequest: Rea
 
 					const reactivateUserByUIDWhere: QueryType<UserAuth> = {
 						uid,
-						role: 'blocked',
+						role: 'baned',
 					}
 
 					const reactivateUserByUIDUpdate: UpdateType<UserAuth> = {
@@ -2507,7 +2507,7 @@ export const reactivateUserByUIDService = async (reactivateUserByUIDRequest: Rea
  * @param adminToken 管理员的 Token
  * @returns 获取所有被封禁用户的信息的请求响应
  */
-export const getBlockedUserService = async (adminUid: number, adminToken: string): Promise<GetBlockedUserResponseDto> => {
+export const getBannedUserService = async (adminUid: number, adminToken: string): Promise<GetBannedUserResponseDto> => {
 	try {
 		if (await checkUserToken(adminUid, adminToken)) {
 			const isAdmin = await checkUserRoleService(adminUid, 'admin')
@@ -2515,10 +2515,10 @@ export const getBlockedUserService = async (adminUid: number, adminToken: string
 				const { collectionName: userAuthCollectionName, schemaInstance: userAuthSchemaInstance } = UserAuthSchema
 
 				// TODO: 下方这个 Aggregate 只适用于被封禁用户的搜索
-				const blockedUserAggregateProps: PipelineStage[] = [
+				const banedUserAggregateProps: PipelineStage[] = [
 					{
 						$match: {
-							role: 'blocked',
+							role: 'baned',
 						},
 					},
 					{
@@ -2552,7 +2552,7 @@ export const getBlockedUserService = async (adminUid: number, adminToken: string
 				]
 
 				try {
-					const userResult = await selectDataByAggregateFromMongoDB(userAuthSchemaInstance, userAuthCollectionName, blockedUserAggregateProps)
+					const userResult = await selectDataByAggregateFromMongoDB(userAuthSchemaInstance, userAuthCollectionName, banedUserAggregateProps)
 					if (userResult && userResult.success) {
 						const userInfo = userResult?.result
 						if (userInfo?.length > 0) {
@@ -4403,16 +4403,16 @@ const checkCheckUsernameRequest = (checkUsernameRequest: CheckUsernameRequestDto
 
 /**
  * 检查封禁用户的请求载荷
- * @param blockUserByUIDRequest 封禁用户的请求载荷
+ * @param banUserByUIDRequest 封禁用户的请求载荷
  * @returns 检查结果，合法返回 true，不合法返回 false
  */
-const checkBlockUserByUIDRequest = (blockUserByUIDRequest: BlockUserByUIDRequestDto): boolean => {
-	return (blockUserByUIDRequest.criminalUid !== null && blockUserByUIDRequest.criminalUid !== undefined)
+const checkBanUserByUIDRequest = (banUserByUIDRequest: BanUserByUIDRequestDto): boolean => {
+	return (banUserByUIDRequest.criminalUid !== null && banUserByUIDRequest.criminalUid !== undefined)
 }
 
 /**
  * 检查重新激活用户的请求载荷
- * @param blockUserByUIDRequest 重新激活用户的请求载荷
+ * @param banUserByUIDRequest 重新激活用户的请求载荷
  * @returns 检查结果，合法返回 true，不合法返回 false
  */
 const checkReactivateUserByUIDRequest = (reactivateUserByUIDRequest: ReactivateUserByUIDRequestDto): boolean => {

```

YasaitoriOvO committed:
:art: 重命名封禁相关函数

YasaitoriOvO committed:
✨ 添加封禁用户功能

YasaitoriOvO committed:
✨ 添加隐藏用户功能

YasaitoriOvO committed:
✨ 添加封禁与解封功能

YasaitoriOvO committed:
:art: 重构封禁数据库结构

YasaitoriOvO committed:
:art: QoL Update

cfdxkk committed:
Merge remote-tracking branch 'origin/feature-20250105-Feed' into merge-fix-20250405

cfdxkk committed:
:twisted_rightwards_arrows: Fix merge conflict

cfdxkk committed:
🐛 Fix unable unfollowing

YasaitoriOvO committed:
✨ 添加封禁相关内容的 API

YasaitoriOvO committed:
Merge branch 'feature-20250105-Feed' of https://github.com/KIRAKIRA-DOUGA/KIRAKIRA-Rosales into feature-20250105-Feed

YasaitoriOvO committed:
:art: QoL Update

YasaitoriOvO committed:
✅ 修复并测试封禁相关内容 & 🎨 QoL Update

YasaitoriOvO committed:
:art: QoL Update

YasaitoriOvO committed:
🚚 封禁 -> 屏蔽

YasaitoriOvO committed:
♻️ 🚧 重构封禁逻辑

YasaitoriOvO committed:
♻️ 重构屏蔽函数

YasaitoriOvO committed:
♻️ 重构"取消屏蔽用户"函数

YasaitoriOvO committed:
♻️ 重构取消屏蔽相关函数

YasaitoriOvO committed:
✨ 新增获取黑名单的函数

YasaitoriOvO committed:
✨ 新增检测的函数

YasaitoriOvO committed:
✨ 新增黑名单相关接口

YasaitoriOvO committed:
✨ 增加上限

YasaitoriOvO committed:
📝 更新 Readme 及任务

YasaitoriOvO committed:
📝 Update Readme

YasaitoriOvO committed:
✅ 测试并修复 Bug

cfdxkk committed:
📝 Update comments

cfdxkk committed:
📝 Update comments

cfdxkk committed:
:art: Remove useless import in BrowsingHistoryService.ts

cfdxkk commented on src/service/BlockService.ts:
```diff
@@ -0,0 +1,1349 @@
+		}
+
+		const insertResult = await insertData2MongoDB<BlockListSchemaType>(blockListData, blockUserSchemaInstance, blockUserCollectionName)
+		if (!insertResult) {
```
`insertResult` 的数据类型如下：
``` typescript
/**
 * 数据操作的结果列表（结果为对象数组）
 */
export type DbPoolResultsType<T> = {
	/** 操作是否成功 */
	success: boolean;
	/** 附加消息 */
	message: string;
	/** 错误信息（如果有的话） */
	error?: unknown;
	/** 数据操作的结果数组（如果有的话） */
	result?: T[];
}
```
如需判定数据库插入操作是否成功，请使用：
``` typescript
if (!insertResult.success) {
	console.error('ERROR', '......')
}
```

cfdxkk commented on src/service/BlockService.ts:
```diff
@@ -0,0 +1,1349 @@
+		return { success: false, message: '删除正则表达式失败，未知错误' }
+	}
+}
+
```
上方多个新增和删除的接口也许可以合并为一个新增和一个删除。

cfdxkk replied:
update: 已放弃该修改

cfdxkk committed:
✨ 动态分组相关接口适配 RBAC

cfdxkk committed:
🐛 修复插入失败的查询结果会被判定为成功的问题。:art: 一般性代码风格修复。

YasaitoriOvO committed:
🎨 mute -> hide
改了又改 最终回到原点...

cfdxkk committed:
✨ :recycle: 视频评论接口接入黑名单功能。

cfdxkk committed:
✨ ♻️ 首页视频接口接入黑名单功能。

cfdxkk commented:
这个分支里也修复了前端存储库的这个 issue：https://github.com/KIRAKIRA-DOUGA/KIRAKIRA-Cerasus/issues/242

cfdxkk committed:
✨ ♻️ 视频弹幕接口接入黑名单功能。

cfdxkk committed:
🐛 尽管 HTTP 规范没有禁止 GET 请求携带请求体，但其语义尚未定义，服务端也可能会抛出错误。

cfdxkk committed:
✨ 获取黑名单 TAG 列表

cfdxkk committed:
:lock: 修复 Regex Catastrophic Backtracking（灾难性回溯）漏洞

关于 Catastrophic Backtracking（灾难性回溯）的介绍性文章：https://zh.javascript.info/regexp-catastrophic-backtracking

cfdxkk committed:
:memo: 添加 buildBlockListMongooseFilter 函数的参数与返回值的类型注释

cfdxkk committed:
🎨 修改一下 BlockControllerDto 中类型的顺序

cfdxkk committed:
🔒 修复 Keyword Catastrophic Backtracking（灾难性回溯）漏洞

cfdxkk committed:
✨ 关键词黑名单和正则表达式黑名单，以及其他 QoL 改进

1. 修改单个关键词和正则表达式的长度限制
2. 正则表达式使用模糊匹配
3. 其他 QoL 改进

YasaitoriOvO committed:
🚑 修复未校验用户名/昵称的问题

YasaitoriOvO committed:
🔥 删除实验性功能（扁平模式/直角模式）

cfdxkk committed:
✨ 完善屏蔽与隐藏功能，修复大量问题

cfdxkk committed:
:egg: :beers: add meme
